x/ibc-transfer/types/trace.go

```diff
@@ -0,0 +1,196 @@
+	case dt.Trace != "" && strings.TrimSpace(dt.BaseDenom) == "":
+		return sdkerrors.Wrap(ErrInvalidDenomForTransfer, "denomination cannot be blank")
+	}
+
+	// NOTE: no base denomination validation
+
+	identifiers := strings.Split(dt.Trace, "/")
+	return validateTraceIdentifiers(identifiers)
+}
+
+// Traces defines a wrapper type for a slice of IdentifiedDenomTraces.
+type Traces []DenomTrace
+
+// Validate performs a basic validation of each denomination trace info.
+func (t Traces) Validate() error {
+	seenTraces := make(map[string]bool)
+	for i, trace := range t {
+		hash := trace.Hash().String()
+		if seenTraces[hash] {
+			return fmt.Errorf("duplicated denomination trace with hash  %s", trace.Hash())
+		}
+
+		if err := trace.Validate(); err != nil {
+			return sdkerrors.Wrapf(err, "failed denom trace %d validation", i)
+		}
+		seenTraces[hash] = true
+	}
+	return nil
+}
+
+// ValidatePrefixedDenom checks that the denomination for an IBC fungible token packet denom is correctly prefixed.
+// The function will return no error if the given string follows one of the two formats:
+//
+//  - Prefixed denomination: '{portIDN}/{channelIDN}/.../{portID0}/{channelID0}/baseDenom'
+//  - Unprefixed denomination: 'baseDenom'
+func ValidatePrefixedDenom(denom string) error {
+	denomSplit := strings.Split(denom, "/")
+	if denomSplit[0] == denom && strings.TrimSpace(denom) != "" {
+		// NOTE: no base denomination validation
+		return nil
+	}
+
+	if strings.TrimSpace(denomSplit[len(denomSplit)-1]) == "" {
+		return sdkerrors.Wrap(ErrInvalidDenomForTransfer, "base denomination cannot be blank")
+	}
+
+	identifiers := denomSplit[:len(denomSplit)-1]
+	return validateTraceIdentifiers(identifiers)
+}
+
+// ValidateIBCDenom validates that the given denomination is either:
+//
+//  - A valid base denomination (eg: 'uatom')
+//  - A valid trace prefixed denomination  (eg: '{portIDN}/{channelIDN}/.../{portID0}/{channelID0}/baseDenom')
+//  - A valid fungible token representation (i.e 'ibc/{hash}') per ADR 001 https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-001-coin-source-tracing.md
+func ValidateIBCDenom(denom string) error {
+	denomSplit := strings.Split(denom, "/")
+
+	switch {
+	case strings.TrimSpace(denom) == "",
+		len(denomSplit) == 1 && denomSplit[0] == "ibc",
+		len(denomSplit) == 2 && (denomSplit[0] != "ibc" || strings.TrimSpace(denomSplit[1]) == ""):
+		return sdkerrors.Wrapf(ErrInvalidDenomForTransfer, "denomination should be prefixed with the format 'ibc/{hash(trace + \"/\" + %s)}'", denom)
+
+	case denomSplit[0] == denom && strings.TrimSpace(denom) != "":
+		return sdk.ValidateDenom(denom)
+
+	case len(denomSplit) > 2:
+		return validateTraceIdentifiers(denomSplit[:len(denomSplit)-1])
+	}
+
+	if _, err := ParseHexHash(denomSplit[1]); err != nil {
+		return sdkerrors.Wrapf(ErrInvalidDenomForTransfer, "invalid denom trace hash %s: %s", denomSplit[1], err)
```

Comment:
```suggestion
		return sdkerrors.Wrapf(err, "invalid denom trace hash %s", denomSplit[1])
```
nit: an ibc/hash is technically not used for cross chain transfers